system.c

```diff
@@ -57,6 +75,57 @@ ISR(TIMER2_COMPA_vect)
   masterclock++;
 }
 
+/* KEYME SPECIFIC START */
+void init_ADC(){
+  ADCSRB = 0;
+  ADCSRA = 0;
+  ADMUX = (1<<REFS0); // Set ADC reference
+  ADCSRA = (1<<ADPS2)|(1<<ADPS0); // Enable prescaler of 32x
+  ADCSRA |= (1<<ADIE); // Enable ADC interrupt
+  ADCSRA |= (1<<ADEN); // Enable ADC
+  ADCSRA |= (1<<ADATE); // Enable AutoTriggering
+  ADCSRB = (1<<ADTS2)|(1<<ADTS0); // Start conversion on Timer1 CTC
+}
+
+/* This performs a low pass filter on ADC values to reduce impact of noise on a final
+   value */
+uint32_t lpf(float beta, uint32_t adc, uint32_t final_prev){
+  float result = (float)final_prev - (beta * ((float)final_prev - (float)adc));
+  return (uint32_t)result;
+}
+
+/* The ADC completion Interrupt: This interrupt occurs once ADC completes its conversion
+   of target. The ADC is set to AutoTrigger, which means the conversion will begin at certain
+   intervals. In this case, the conversion begins when the TIMER1_COMPA_vector signals an
+   interrupt. Voltage values are stored in an array. Values in Array are printed on request
+   in report_voltage(). */
+ISR(ADC_vect){
+  // These bits are flipped to 0 while inside TIMER1_COMPA_vector interrupt and to 1 once we
+  // exit the interrupt. Since there is no ISR for this, we must manually flip them to signal
+  // that we have exited this interrupt.
+  TIFR1 |= (1<<OCF1A)|(1<<OCF1B);
+  
+  float beta_LPF = 0.25; // coefficient for low pass filter
+  // Final conversion is a 10 bit value stored in ADC
+  voltage_result[voltage_result_index] = lpf(beta_LPF, (uint32_t)ADC,
```

Comment:
The Atmega2560 is an 8-bit microcontroller. The final ADC value is reported over two 8-bit registers (`ADCL `and `ADCH`). The value described by "`ADC`" is a 16 bit value that consists of both `ADCL `and `ADCH`. I first cast `ADC `to a `uint32_t `to extend the 0's. I cast these values to floats in `lpf ` before any arithmetic to avoid overflow issues (since they are unsigned values). In `request_voltage`, I use the `printInteger `function to report values. Thus I need to cast back to a `uint32_t `once completing the arithmetic in `lpf`.